pennylane/interfaces/execution.py

```diff
@@ -363,53 +446,63 @@ def cost_fn(params, x):
-    else:
-        batch_fn = lambda res: res  # pragma: no cover
-
     if isinstance(cache, bool) and cache:
         # cache=True: create a LRUCache object
         cache = LRUCache(maxsize=cachesize)
         setattr(cache, "_persistent_cache", False)
 
-    batch_execute = set_shots(device, override_shots)(device.batch_execute)
+    if new_device_interface:
+        batch_execute = device.execute
+    else:
+        batch_execute = set_shots(device, override_shots)(device.batch_execute)
 
-    if expand_fn == "device":
-        expand_fn = lambda tape: device.expand_fn(tape, max_expansion=max_expansion)
+    expand_fn = _preprocess_expand_fn(expand_fn, device, max_expansion)
 
-    if gradient_fn is None:
-        # don't unwrap if it's an interface device
-        if "passthru_interface" in device.capabilities() or device.short_name == "default.mixed":
-            return batch_fn(
-                qml.interfaces.cache_execute(
-                    batch_execute, cache, return_tuple=False, expand_fn=expand_fn
-                )(tapes)
-            )
-        unwrapped_tapes = tuple(qml.transforms.convert_to_numpy_parameters(t) for t in tapes)
-        res = qml.interfaces.cache_execute(
-            batch_execute, cache, return_tuple=False, expand_fn=expand_fn
-        )(unwrapped_tapes)
+    #### Executing the configured setup #####
 
-        return batch_fn(res)
+    tapes, batch_fn = _batch_transform(
+        tapes, device, config, override_shots, device_batch_transform
+    )
 
-    if gradient_fn == "backprop" or interface is None:
-        return batch_fn(
-            qml.interfaces.cache_execute(
-                batch_execute, cache, return_tuple=False, expand_fn=expand_fn
-            )(tapes)
+    # Exiting early if we do not need to deal with an interface boundary
+    no_interface_boundary_required = interface is None or gradient_fn in {None, "backprop"}
+    if no_interface_boundary_required:
+
+        device_supports_interface_data = (
+            new_device_interface
+            or config.interface is None
+            or gradient_fn == "backprop"
+            or device.short_name == "default.mixed"
+            or "passthru_interface" in device.capabilities()
         )
+        if not device_supports_interface_data:
+            tapes = tuple(qml.transforms.convert_to_numpy_parameters(t) for t in tapes)
+
+        # use qml.interfaces so that mocker can spy on it during testing
+        cached_execute_fn = qml.interfaces.cache_execute(
+            batch_execute,
+            cache,
+            expand_fn=expand_fn,
+            return_tuple=False,
+            pass_kwargs=new_device_interface,
+        )
+        results = cached_execute_fn(tapes, execution_config=config)
+        return batch_fn(results)
 
     # the default execution function is batch_execute
+    # use qml.interfaces so that mocker can spy on it during testing

```

Comment:
does spy not work if you import the function directly? might need some extra finagling but that's surprising to me. I usually use `unittest.mock` so it's new territory for me.

Reply:
spying is black magic to me.  I tried to just `cache_execute` and a whole bunch of tests started failing, so I reverted the change and put a reminder to not try that again.

Unless someone wants to actually fix it, but I'm not sure its worth it.